File: Maximum Rectangle/Maximum Rectangle.cpp
class Solution {
public:
    int maximalRectangle(vector<vector<char> > &matrix) {
        int max = 0;
        // Row forms the length
        // Find the maximum height * length
        for(int row = 0; row < matrix.size(); ++row){
            int length = 0;
            std::vector<int> minHeight;
            for(int col = 0; col < matrix[0].size(); ++col){
                if(matrix[row][col] == '1'){
                    ++length;
                    int height = 1;
                    
                    while(row + height < matrix.size()){
                        if(matrix[row + height][col] == '1'){
                            ++height;
                        }
                        else{
                            break;
                        }
                    }
                    
                    minHeight.push_back(height);
                    for(int i = 0; i < minHeight.size(); ++i){
                        minHeight[i] = (minHeight[i] < height) ? minHeight[i] : height;
                        // As length increases, the height is likely to decrease
                        int l = length - i;
                        int area = minHeight[i] * l;
                        max = (max > area) ? max : area;
                    }
                    
                }
                else{
                    length = 0;
                    minHeight.clear();
                }
            }
        }
        
        return max;
    }
};

File: Maximum Rectangle/Maximum Rectangle.java
public class Solution {
    public int maximalRectangle(char[][] matrix) {
        int rowSize = matrix.length;
        if(rowSize == 0)                        return 0;
        int colSize = matrix[0].length;
        if(colSize == 0)                        return 0;
        
        int maxArea = 0;
        for(int i = 0; i < rowSize; ++i){
            List<Integer> height = new ArrayList<Integer>();
            
            for(int j = 0; j < colSize; ++j){
                if(matrix[i][j] == '1'){
                    int currHeight = 1;
                    while(currHeight + i < rowSize){
                        if(matrix[currHeight + i][j] == '0'){
                            break;
                        }
                        else{
                            ++currHeight;
                        }
                    }
                    
                    height.add(currHeight);
                    for(int p = 0; p < height.size(); ++p){
                        int len = height.size() - p;
                        height.set(p, Math.min(height.get(p), currHeight));
                        int h = height.get(p);
                        maxArea = Math.max(maxArea, h * len);
                    }
                }
                else{
                    height.clear();
                }
            }
        }
        
        return maxArea;
    }
}
